### PR TITLE
API refactors for ease of use (and yarn!)

### PR DESCRIPTION
**Pull request by alexjpwalker**
## What is the goal of this PR?

We've added an `index.ts` file so that symbols are importable from "typedb-client" as opposed to, say, "typedb-client/api/connection/TypeDBTransaction".

Also, we've added casting methods to Concept classes (e.g: `asThingType`). These methods perform safe type casting.

We've aligned the class/namespace structure fully with client-java - previously, for example, `AttributeType.Boolean.Remote` was flattened to `AttributeType.RemoteBoolean`. This is no longer the case.

We've found `yarn` is faster and more deterministic than `npm` for installing dependencies, so we've updated the build system to use `yarn`. This does not affect users.

(Also, the generated lockfile is **much** smaller!) 

## What are the changes implemented in this PR?

- Shorten import paths
- Add Concept type cast methods
- Flatten class structures
- Use `yarn` instead of `npm` to install dependencies

### CODE REVIEW

**Comment by flyingsilverfin** [comment]
plz add milestone and make it not read :P

**Comment by flyingsilverfin** [comment]
plz comment findings on `..._` typing in this PR for future reference to save time

**Review by flyingsilverfin — APPROVED** [review]

File: concept/ConceptImpl.ts [hunk 1]
```diff
@@ -71,20 +85,107 @@ export abstract class ConceptImpl implements Concept {
...
 }
 
 export namespace ConceptImpl {
 
-    export abstract class Remote extends ConceptImpl implements RemoteConcept {
+    export abstract class Remote extends ConceptImpl implements Concept.Remote {
+
+        private readonly _transaction: TypeDBTransaction.Extended;
+
+        protected constructor(transaction: TypeDBTransaction.Extended, ..._: any) {
```
Comment (alexjpwalker) [inline concept/ConceptImpl.ts:137]: One unfortunate result of the "hack" where we enable class nesting by having a `namespace` and `class` of the same name, is that TypeScript believes that the constructors of nested classes must be compatible with each other - as if they were regular methods.

Here we have `ConceptImpl.Remote.constructor` conflicting with `ThingImpl.Remote.constructor`. If we were to rename the namespace to, say, `ConceptImpl_1`, then we wouldn't need the `..._` varargs.

The effect of `..._: any` is to say that this constructor can take additional arguments. This makes its signature "compatible" with the signature of `ThingImpl.Remote.constructor` (which additionally has `iid` and `isInferred` not present in `ConceptImpl.Remote`)

File: api/concept/thing/Attribute.ts [hunk 2]
```diff
@@ -19,20 +19,24 @@
...
 
-export interface Attribute<T extends AttributeType.ValueClass> extends Thing {
+export interface Attribute extends Thing {
 
-    asRemote(transaction: TypeDBTransaction): Attribute.Remote<T>;
+    asRemote(transaction: TypeDBTransaction): Attribute.Remote;
 
     getType(): AttributeType;
 
-    getValue(): T;
```
[the remark this answers is not in the document]
Comment (alexjpwalker) [reply api/concept/thing/Attribute.ts:34]: Reinstated as `getValue(): string | number | boolean | Date`

File: api/concept/type/RelationType.ts [hunk 3]
```diff
@@ -19,39 +19,66 @@
...
 import {TypeDBTransaction} from "../../connection/TypeDBTransaction";
 import {Stream} from "../../../common/util/Stream";
 import {RoleType} from "./RoleType";
 import {Relation} from "../thing/Relation";
+import { Type } from "./Type";
+import { EntityType } from "./EntityType";
+import { Thing } from "../thing/Thing";
+import { Entity } from "../thing/Entity";
+import { Attribute } from "../thing/Attribute";
+import { AttributeType } from "./AttributeType";
```
[the remark this answers is not in the document]
Comment (alexjpwalker) [reply api/concept/type/RelationType.ts]: Reformatted whole project with IntelliJ sorting imports by path, symbols imported by name, and padding imported symbols with spaces

**Pull request merged**